lib/src/rules/literal_only_boolean_expression.dart

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.literal_only_boolean_expression;
+
+import 'package:linter/src/linter.dart';
+import 'package:analyzer/analyzer.dart';
+
+const _desc = r'Conditions should not unconditionally evaluate to "TRUE" or to "FALSE"';
+
+const _details = r'''
+
+**DON'T** test for conditions composed only by literals, since the value can be
+inferred at compile time.
+Conditional statements using a condition which cannot be anything but FALSE have
+the effect of making blocks of code non-functional. If the condition cannot
+evaluate to anything but TRUE, the conditional statement is completely
+redundant, and makes the code less readable.
+It is quite likely that the code does not match the programmer's intent.
+Either the condition should be removed or it should be updated so that it does
+not always evaluate to TRUE or FALSE.
+
+**BAD:**
+```
+void bad() {
+  if (true) {} // LINT
+}
+```
+
+**BAD:**
+```
+void bad() {
+  if (true && 1 != 0) {} // LINT
+}
+```
+
+**BAD:**
+```
+void bad() {
+  if (1 != 0 && true) {} // LINT
+}
+```
+
+**BAD:**
+```
+void bad() {
+  if (1 < 0 && true) {} // LINT
+}
+```
+
+**BAD:**
+```
+void bad() {
+  if (true && false) {} // LINT
+}
+```
+
+**BAD:**
+```
+void bad() {
+  if (1 != 0) {} // LINT
+}
+```
+
+**BAD:**
+```
+void bad() {
+  if (true && 1 != 0 || 3 < 4) {} // LINT
+}
+```
+
+**BAD:**
+```
+void bad() {
+  if (1 != 0 || 3 < 4 && true) {} // LINT
+}
+```
+
+''';
+
+class LiteralOnlyBooleanExpression extends LintRule {
+  _Visitor _visitor;
+
+  LiteralOnlyBooleanExpression() : super(
+      name: 'literal_only_boolean_expression',
+      description: _desc,
+      details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  visitWhileStatement(WhileStatement node) {
+    if (_onlyLiterals(node.condition)) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  visitIfStatement(IfStatement node) {
+    if (_onlyLiterals(node.condition)) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  visitForStatement(ForStatement node) {
+    if (_onlyLiterals(node.condition)) {
+      rule.reportLint(node);
+    }
+  }
+}
+
+bool _onlyLiterals(Expression expression) {
```

Comment:
Do we want to special case `??`? For example, `3 ?? m()` might want to return `true`, while `null ?? m()` would return `false`.